Base/usr/share/shell/completion/builtin.sh

```diff
@@ -74,3 +74,17 @@ _complete_kill() {
     __complete_job_spec $*[-1]
   }
 }
+
+_complete_cd() {
+    if test $*[-1] = '--' {
+        invariant_offset=0
+        dir_glob=(.*/ */)
+    } else {
+        invariant_offset=${length "$*[-1]"}
+        dir_glob="$*[-1]*/"
+    }
+
+    for $(glob $dir_glob) {
```

Comment:
This will reevaluate file paths as a glob, so a dir named `..*/` (I know, dumb example) will give you duplicate entries.
Instead, how about
```suggestion
    if test $*[-1] = '--' {
        invariant_offset=0
        results=(.*/ */)
    } else {
        invariant_offset=${length "$*[-1]"}
        results=$(glob "$*[-1]*/")
    }
    for $results {
```